### PR TITLE
openldap: handle connect phase with a state machine

### PR DESCRIPTION
**Pull request by monnerat**
This is the first step of my work on openldap.
This allows future implementation of explicit TLS and SASL authentication.

### CODE REVIEW

File: lib/openldap.c [hunk 1]
```diff
@@ -205,8 +259,68 @@ static CURLcode oldap_setup_connection(struct Curl_easy *data,
...
+    passwd.bv_len = strlen(passwd.bv_val);
+  }
+
+  rc = ldap_sasl_bind(li->ld, binddn, LDAP_SASL_SIMPLE, &passwd,
+                      NULL, NULL, &li->msgid);
+  if(rc == LDAP_SUCCESS)
+    state(data, newstate);
+  else
+    result = oldap_map_error(rc, conn->bits.user_passwd?
+                                 CURLE_LOGIN_DENIED: CURLE_LDAP_CANNOT_BIND);
```
Comment (bagder) [inline lib/openldap.c:287]: unexpected indent level ^

File: lib/openldap.c [hunk 2]
```diff
@@ -205,8 +259,68 @@ static CURLcode oldap_setup_connection(struct Curl_easy *data,
...
+  return result;
+}
+
 #ifdef USE_SSL
 static Sockbuf_IO ldapsb_tls;
```
Comment (bagder) [inline lib/openldap.c:292]: I know you didn't add this, but this is really bad isn't it? A static variable like this seems like a potential breaker of many things if we do many LDAP parallel transfers, right?

**Review by bagder — APPROVED** [review]

File: lib/openldap.c (hunk 2 above)
Comment (bagder) [reply lib/openldap.c:292]: Strike that. This struct is just full of function pointers and should be fine to reuse.

File: lib/openldap.c (hunk 1 above)
Comment (monnerat) [reply lib/openldap.c]: Fixed.

File: lib/openldap.c (hunk 2 above)
Comment (monnerat) [reply lib/openldap.c:293]: I don't like this very much either. But as you say, it should'nt harm.

**Comment by bagder** [comment]
Thanks!

**Comment by monnerat** [comment]
Thanks for pulling.

**Pull request closed without merging**